timemachine/cpp/src/kernels/k_sampling.cu

```diff
@@ -38,6 +38,67 @@ template void __global__ k_setup_gumbel_max_trick<double>(
     const double *__restrict__ gumbel_noise,
     double *__restrict__ prepared_gumbel);
 
+template <typename RealType>
+void __global__ k_setup_gumbel_max_trick_with_offset(
+    const int num_segments,
+    const int total_values,
+    const int max_offset,
+    const int *__restrict__ noise_offset,     // [1]
+    const int *__restrict__ segment_offsets,  // [num_segments]
+    const RealType *__restrict__ log_weights, // [total_values]
+    const RealType *__restrict__ gumbel_noise,
+    RealType *__restrict__ prepared_gumbel) {
+    const int segment_idx = blockIdx.y;
+
+    if (segment_idx >= num_segments) {
+        return;
+    }
+
+    const int rand_offset = noise_offset[0];
+    if (rand_offset + segment_idx >= max_offset) {
+        return;
+    }
+
+    const int segment_start = segment_offsets[segment_idx];
+    const int N = segment_offsets[segment_idx + 1] - segment_start;
+    // In the case of the offset the values per segment need to match up to compute gumbel offset correctly.
+    assert(total_values % N == 0);
+
+    const int gumbel_offset = rand_offset * N + segment_start;
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    while (idx < N) {
+        const RealType weight = log_weights[idx];
+        assert(!isnan(weight));
+
+        // If the idx in the batch segments is beyond the max offset return a negative infinity to avoid sampling the weight
+        const RealType gumbel_rand = -log(-log(gumbel_noise[gumbel_offset + idx]));
```

Comment:
sanity checking:

`gumbel_offset + idx = rand_offset * N + segment_start + idx`

is always going to be less than `N * num_proposals_per_move`?

The while statement guarantees that `idx < N` - but less clear on the `rand_offset * N + segment_start` part

Reply:
In the case of BiasedDeletion, the `rand_offset` represents the number the number of segments that have been computed previously. Note that this is likely to change for TIBD, but don't want to muddy the waters in this PR where the following logic is correct.

The `gumbel_noise` array is constructed with length `num_target_mols_ * num_proposals_per_move_` where `num_target_mols == N` in this kernel. 

So `rand_offset * num_target_mols (N in this kernel)` as long as the value of `rand_offset + segment_idx < num_proposals_per_move`, where `segment_idx` is the segment to evaluate, is safe and will result in at most `num_target_mols * (num_proposals_per_move - 1)`. We have a check for this here: https://github.com/proteneer/timemachine/blob/feature/batch-bd-mover/timemachine/cpp/src/kernels/k_sampling.cu#L58 where `max_offset == num_proposals_per_move`. Leaving us with the `segment_start + idx` component.

The `segment_offsets` is expected to be a a set of offsets into the arrays that are to be sampled, with `segment_offsets[0] == 0` and `segment_offsets[-1] == num_target_mols * num_segments`. In this case we take the delta from `segment_offsets[i + 1] - segment_offsets[i]` giving us the number of values in the segment. We start from `segment_offsets[i]` and iterate until we hit `segment_offsets[i+1] - 1` which at most be `num_target_mols - 1` which when added to the `rand_offset * N` will at most be `(num_target_mols * num_proposals_per_move) - 1` which is the expected final value this kernel should achieve.

Reply:
Added a test with only a single mol which would catch if we were overrunning the buffer (its running under `compute-sanitizer --tool memcheck`)